File: backend/Testing/Browser/UserPageTest.cs
using Testing.Browser.Base;
using Testing.Browser.Page;
using Testing.Browser.Page.External;

namespace Testing.Browser;

[Trait("Category", "Integration")]
public class UserPageTest : PageTest
{
    public override async Task InitializeAsync()
    {
        await base.InitializeAsync();
    }

    [Fact]
    public async Task CanUpdateAccountInfo()
    {
        await using var userDashboardPage = await RegisterUser("Test: Edit account - update", $"{Guid.NewGuid()}@mailinator.com", "test_edit_account_update");
        var userPage = await new UserAccountSettingsPage(Page).Goto();
        await userPage.FillDisplayName("Test: Edit account - update - changed");
        await userPage.FillEmail($"{Guid.NewGuid()}_changed@test.com");
        await userPage.ClickSave();
        await Page.GetByText("Your account has been updated.").WaitForAsync();
        await userPage.EmailVerificationAlert.AssertRequestedToChange();
    }

    [Fact]
    public async Task DisplaysFormErrorsOnInvalidData()
    {
        await using var userDashboardPage = await RegisterUser("Test: Edit account - errors", $"{Guid.NewGuid()}@mailinator.com", "test_edit_account_errors");
        var userPage = await new UserAccountSettingsPage(Page).Goto();
        await userPage.FillDisplayName("");
        await userPage.FillEmail("");
        await userPage.ClickSave();
        await Expect(Page.GetByText("Invalid email")).ToBeVisibleAsync();
    }

    [Fact]
    public async Task CanResetPassword()
    {
        var mailinatorId = Guid.NewGuid().ToString();
        await using var userDashboardPage = await RegisterUser("Test: Edit account - reset password", $"{mailinatorId}@mailinator.com", "test_edit_account_reset_password");
        var user = userDashboardPage.User;
        var userPage = await new UserAccountSettingsPage(Page).Goto();
        var resetPasswordPage = await userPage.ClickResetPassword();
        var newPassword = "test_edit_account_reset_password_changed";
        await resetPasswordPage.FillForm(newPassword);
        await resetPasswordPage.Submit();

        var loginPage = await Logout();
        await loginPage.FillForm(userDashboardPage.User.Email, newPassword);
        await loginPage.Submit();
        await new UserDashboardPage(Page).WaitFor();

        //verify password changed email was received
        var inboxPage = await MailInboxPage.Get(Page, mailinatorId).Goto();
        var emailPage = await inboxPage.OpenEmail();
        await Expect(emailPage.Page.GetByText("Your password was changed").First).ToBeVisibleAsync();
    }
}

File: frontend/tests/userPage.test.ts
import { loginAs, logout } from './utils/authHelpers';

import { EmailSubjects } from './pages/mailPages';
import { UserAccountSettingsPage } from './pages/userAccountSettingsPage';
import { UserDashboardPage } from './pages/userDashboardPage';
import { expect } from '@playwright/test';
import { getInbox } from './utils/mailboxHelpers';
import { randomUUID } from 'crypto';
import { test } from './fixtures';

test('can update account info', async ({ page, tempUser }) => {
  await loginAs(page.request, tempUser.email, tempUser.password);
  const userPage = await new UserAccountSettingsPage(page).goto();
  const newUuid = randomUUID();
  await userPage.fillDisplayName('Test: Edit account - update - changed');
  await userPage.fillEmail(`${newUuid}_changed@test.com`);
  await userPage.clickSave();
  await page.getByText('Your account has been updated.').waitFor();
  await userPage.emailVerificationAlert.assertRequestedToChange();
});

test('display form errors on invalid data', async ({ page, tempUser }) => {
  await loginAs(page.request, tempUser.email, tempUser.password);
  const userPage = await new UserAccountSettingsPage(page).goto();
  await userPage.fillDisplayName('');
  await userPage.fillEmail('');
  await userPage.clickSave();
  await expect(page.getByText('Invalid email')).toBeVisible();
});

test('can reset password', async ({ page, tempUser }) => {
  test.slow(); // email checking can be slow

  const newPassword = 'test_edit_account_reset_password_changed';
  expect(tempUser.password).not.toBe(newPassword);
  await loginAs(page.request, tempUser.email, tempUser.password);
  const userPage = await new UserAccountSettingsPage(page).goto();
  const resetPasswordPage = await userPage.clickResetPassword();
  await resetPasswordPage.fillForm(newPassword);
  await resetPasswordPage.submit();

  const loginPage = await logout(page);
  await loginPage.fillForm(tempUser.email, newPassword);
  await loginPage.submit();

  await new UserDashboardPage(loginPage.page).waitFor();

  // Verify password changed email was received
  const inboxPage = await getInbox(page, tempUser.mailinatorId).goto();
  const emailPage = await inboxPage.openEmail(EmailSubjects.PasswordChanged);
  await expect(emailPage.page.getByText('Your password was changed').first()).toBeVisible();
});
